lazyflow/classifiers/lazyflowClassifier.py

```diff
@@ -260,12 +260,166 @@ def serialize_hdf5(self, h5py_group):
+        2) override the proper instance methods
+        3) have a VERSION class member
+        """
+        if cls is LazyflowPixelwiseClassifierFactoryABC:
+            is_subclass = LazyflowPixelwiseClassifierFactoryABC in C.__mro__
+            is_subclass &= _has_attributes(C, ["create_and_train_pixelwise", "description", "get_halo_shape"])
+            is_subclass &= "VERSION" in C.__dict__
+            return is_subclass
+        return NotImplemented
+
+    def __eq__(self, other):
+        """
+        Classifier factories must be both copyable and (in)equality comparable.
+        """
+        raise NotImplementedError
+
+    def __ne__(self, other):
+        raise NotImplementedError
+
+    """
+    Defines an interface for "pixel-wise" classifier objects that can be used by the lazyflow classifier operators.
+    A "pixel-wise" classifier expects its input be given as a list of ND feature images (with M feature channels).
+    (It was trained with a list of ND label images, with 1 channel each.)
+
+    Note: It is assumed here that 'channel' is always the last axis of the image.
+
+    (This interface is typically used with classifiers that must generate their own features internally,
+    and thus require the knowledge of the image structure and context around each training/prediction point.)
+    """
+
+    @abc.abstractmethod
+    def predict_probabilities_pixelwise(self, feature_image, roi, axistags=None):
+        """
+        For each pixel in the given feature_image, predict the probabilities that the
+        pixel belongs to each label class the classifier was trained with.
+
+        feature_image: An ND image.  Last axis must be channel.
+        roi: The region of interest (start, stop) within feature_image to predict (e.g. without the halo region)
+             Note: roi parameter should not include channel.
+                   For example, a valid roi for a zyxc image could be ((0,0,0), (10,20,30))
+        axistags: Optional.  A vigra.AxisTags object describing the feature_image.
+
+        Returns: A multi-channel image (each channel corresponds to a different label class).
+                 The result image size is determined by the roi parameter.
+        """
+        raise NotImplementedError
+
+    @abc.abstractproperty
+    def known_classes(self):
+        """
+        Returns the list of label classes known to this classifier (i.e. the classes it was trained with).
+        """
+        raise NotImplementedError
+
+    @abc.abstractproperty
+    def feature_count(self):
+        """
+        Return the number of features used to train this classifier.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def get_halo_shape(self, data_axes="zyxc"):
+        """
+        Same as LazyflowPixelwiseClassifierFactoryABC.get_halo_shape().  See that function for details.
+        """
+        raise NotImplementedError
+
+    @classmethod
+    def __subclasshook__(cls, C):
```

Comment:
Are subclasshooks really necessary in these cases? Seems like a design problem...